### PR TITLE
Object migration

### PR DESCRIPTION
**Pull request by psarram**
- Migrates Fixed Position objects to Display Layout objects.
- Migrates Display Layouts to new format.

### CODE REVIEW

File: src/api/objects/ObjectAPI.js [hunk 1]
```diff
@@ -215,6 +227,188 @@ define([
         return utils.makeKeyString(identifier);
     };
 
+    let migrations = [
```
Comment (akhenry) [inline src/api/objects/ObjectAPI.js]: Maybe we should move these into separate source files?

File: src/api/objects/ObjectAPI.js [hunk 2]
```diff
@@ -215,6 +227,188 @@ define([
...
+                            .then(object => {
+                                telemetryObjects[element.id] = object;
+                            });
+                    }
+                });
+
+                return Promise.all(promises)
+                    .then(function () {
+                        newLayoutObject.configuration.items =
+                            migrateFixedPositionConfigurataion(elements, telemetryObjects);
```
Comment (akhenry) [inline src/api/objects/ObjectAPI.js]: Small typo here

File: src/api/objects/ObjectAPI.js [hunk 3]
```diff
@@ -215,6 +227,188 @@ define([
...
+        }
+    ];
+
+    function needsMigration(domainObject) {
+        return migrations.some(m => m.check(domainObject));
+    }
+
+    function migrateObject(domainObject) {
+        return migrations.filter(m => m.check(domainObject))
+            .reduce((o, m) => "").migrate(domainObject);
```
Comment (akhenry) [inline src/api/objects/ObjectAPI.js]: Could you help me understand the role of 'reduce' here? I would expect the filter on migrations to only return one migration, in which case couldn't we just use [0]?

File: src/api/objects/ObjectAPI.js [hunk 4]
```diff
@@ -215,6 +227,188 @@ define([
...
+            && domainObject.type !== 'example.imagery') {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    function migrateDisplayLayout(domainObject, childObjects) {
+        const DEFAULT_GRID_SIZE = [32, 32];
+        let migratedObject = {...domainObject};
```
Comment (akhenry) [inline src/api/objects/ObjectAPI.js]: The spread operator for object literals is unfortunately not supported by all browsers yet. Maybe we could use `Object.assign({}, domainObject)` instead?

Comment (psarram) [reply src/api/objects/ObjectAPI.js]: There are other places in the code base where the spread operator is used. But that's a fair point. I'll change it to Object.assign().

File: src/api/objects/ObjectAPI.js (hunk 2 above)
Comment (psarram) [reply src/api/objects/ObjectAPI.js]: Good catch!

File: src/api/objects/ObjectAPI.js (hunk 3 above)
Comment (psarram) [reply src/api/objects/ObjectAPI.js]: You are right, reduce is unnecessary here.

File: src/api/objects/ObjectAPI.js [hunk 5]
```diff
@@ -154,7 +156,17 @@ define([
             throw new Error('Provider does not support get!');
         }
 
-        return provider.get(identifier);
+        return provider.get(identifier)
```
[the remark this answers is not in the document]
Comment (psarram) [reply src/api/objects/ObjectAPI.js]: I like the idea. Will define a plugin instead.

File: src/api/objects/ObjectAPI.js (hunk 1 above)
Comment (psarram) [reply src/api/objects/ObjectAPI.js]: I'll consider it when writing the plugin.

**Comment by psarram** [comment]
@akhenry Resolved the initial review feedback. Please let me know what you think of the new changes. Thanks!

File: src/plugins/objectMigration/Migrations.js [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+
+define([
+    'uuid'
+], function (
+    uuid
+) {
+    function Migrations() {
```
Comment (akhenry) [inline src/plugins/objectMigration/Migrations.js]: I think the outer 'Migrations' function is redundant, no? Could we could just return the array of migration definitions?

File: src/plugins/objectMigration/plugin.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ * Open MCT includes source code licensed under additional open source
+ * licenses. See the Open Source Licenses file (LICENSES.md) included with
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+
+import Migrations from './Migrations.js'
+
+export default function () {
+    let migrations = Migrations();
```
Comment (akhenry) [inline src/plugins/objectMigration/plugin.js]: As mentioned above, I don't think it's necessary for this to be a function?

File: src/plugins/objectMigration/Migrations.js (hunk 6 above)
Comment (psarram) [reply src/plugins/objectMigration/Migrations.js]: @akhenry you are right, it's redundant.

File: src/plugins/objectMigration/plugin.js (hunk 7 above)
Comment (psarram) [reply src/plugins/objectMigration/plugin.js]: Yes, removed the redundant function.

**Comment by akhenry** [comment]
Looks good, thanks!

**Pull request merged**